Add ability to right click on map to disable spawn points

ocdi opened the pull request:
This PR is a second take at the task based on feedback by the core team.

This changes the storage of disabled spawns to be just an list of spawn point indexes in the Session. The only method to disable a spawn point is using the right click button. This does reduce the discoverability of this particular feature, however it should support maps with many more spawn points than slots. The interaction toggles the point, which I think is more intuitive than the original selection.

![image](https://user-images.githubusercontent.com/2775804/87995906-10464500-cb34-11ea-8be4-953453d878d0.png)

This also handles the scenario where someone joins and prevents a game from starting if there are insufficient available spawns.

![image](https://user-images.githubusercontent.com/2775804/87995761-bd6c8d80-cb33-11ea-86e5-c105e06a07ed.png)

Closes #18298

ocdi committed:
Right click lobby spawns to disable or remove players.

teinarss commented:
Bug: After fiddling around with disabling/enabling spawn points, right click to disable stopped working.

Discussion: What should happen when there are no spawns available for the players (in the case when some are disabled)? In the lobby and when the game started?

ocdi commented:
@teinarss I suspect the bug you are observing is by trial/error design. This relates to your discussion point, which I did think about and tinker with quite a bit.

I am limiting the number of disabled spawn points such that the players in the lobby are able to have a spot if the game started. The problem is if a new player comes in or you add a bot the number of players exceeds the spawns, which would cause the game to crash (I observed that during testing).

If a new person joins the lobby, the game enters into a state where there are more players than available spawn points. In this case I added checks to the autostart logic to prevent auto-start, the start button is disabled and if the start game command is issued (which should be impossible since start is disabled) it sends a message that insufficient spawns are available and the game can't start. This is shown in the Dune2k lobby screenshot above.

This check also originally had an issue where you couldn't clear a disabled spawn, so I changed it so you can always enable a disabled spawn but it still caps the disabling to free spawns. I just double checked this works if all slots fill up and there are 2 disabled spawns, you can clear both and the game.

ocdi commented:
Actually I think I found the issue you were seeing. The count for enabling spawns (and all those other places) didn't factor in observers. I just pushed a commit excluding them.

teinarss commented:
Still not working at 100%, was able to make one starting location not responding anymore after doing some chaos testing.

ocdi commented:
@teinarss I did another round of testing myself and I found 2 minor issues that are hopefully what you discovered.

* I had copied the selection logic which caused the host to ignore other players spawn selections meaning you could disable a spawn that had been selected
* Players could select a disabled spawn as the disabled spawns weren't excluded from either the dropdown or map selection

The second could happen via map selection or if the dropdown is open when a spawn is disabled, the list doesn't change with spawns being toggled. It only populates the menu when it is opened.

So in addition to these, confirmations of previous behaviour:

* Disabling spawns does not override already selected player spawns
* You can only disable up to current player count
* Emptying the lobby I was able to disable all spawns
* Manually selecting spawns and removing players enabled me to disable all spawns again
* Filling the player list with 100% disabled spawns I was able to re-enable (right-click) all and then left-click select to assign to players
* Only the game host / admin can disable spawns

The last thing I noticed was that the wording still referred to Closed Slot from my original PR, which I updated to say Closed spawn to be consistent with Available spawn.

Hopefully that addresses everything.

phrohdoh commented on OpenRA.Game/OpenRA.Game.csproj:
```diff
@@ -5,7 +5,7 @@
     <Optimize>true</Optimize>
     <UseVSHostingProcess>false</UseVSHostingProcess>
-    <LangVersion>5</LangVersion>
+    <LangVersion>6</LangVersion>
```
@OpenRA/core I suspect we aren't ready for this language version bump

Mailaender replied:
While it doesn't seem to effect the .NET framework version for end users. According to https://docs.microsoft.com/dotnet/csharp/language-reference/compiler-options/langversion-compiler-option C# 6.0 is an unofficial draft at the moment. We shouldn't make use of it as of now.

ocdi replied:
Old habits here, @pchote did raise this in the other PR I was working on, which I fixed and downgraded back to 5. Will do the same here, and for the other related items.

phrohdoh commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -69,6 +71,9 @@ public static Session Deserialize(string data)
 							var s = Slot.Deserialize(node.Value);
 							session.Slots.Add(s.PlayerReference, s);
 							break;
+						case nameof(DisabledSpawns):
```
We'll most likely want to use a string literal, like the other cases.
See my comment regarding the `LangVersion` change in `OpenRA.Game.csproj`.

phrohdoh commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -44,6 +44,7 @@ public class LobbyCommands : ServerTrait, IInterpretCommand, INotifyServerStart,
 			{ "faction", Faction },
 			{ "team", Team },
 			{ "spawn", Spawn },
+			{ "toggle-spawn", ToggleSpawn },
```
```suggestion
			{ "toggle_spawn", ToggleSpawn },
```

An underscore, instead of a dash, is consistent with the existing keys.

phrohdoh commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -760,6 +771,23 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
 
+		static bool ToggleSpawn(S server, Connection conn, Session.Client client, string s)
+		{
+			if (!client.IsAdmin) return true;
```
Non-admins can toggle spawns?

ocdi replied:
This particular code path isn't called because the command never gets issued, having said that I was not fully aware of the intent of the bool. I have flipped it to return false. The answer is no they can't.

ocdi commented:
Thanks for the reviews, I addressed the feedback and squashed it to a single commit, rebasing on `bleed`.

teinarss commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -759,6 +770,23 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
 
+		static bool ToggleSpawn(S server, Connection conn, Session.Client client, string s)
+		{
+			if (!client.IsAdmin) return false;
```
I'm pretty sure that our coding style is that this should be two lines.

teinarss commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -759,6 +770,23 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
+			if (!client.IsAdmin) return false;
+
+			var spawn = Exts.ParseIntegerInvariant(s);
+			if (spawn == 0) return false;
```
same here.

teinarss commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -245,27 +251,55 @@ public SlotDropDownOption(string title, string order, Func<bool> selected)
+			var existing = orderManager.LobbyInfo.Clients.Any(c => selectedSpawn == c.SpawnPoint);
+
+			// can't toggle a spawnpoint for an existing player
+			if (existing) return;
```
And here.

ocdi commented:
@teinarss I addressed those, plus spotted another in `Exts.cs` that I fixed also

teinarss commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -368,7 +368,8 @@ void ConnectionStateChanged(OrderManager om)
 					orderManager.LobbyInfo.Slots.Any(sl => sl.Value.Required && orderManager.LobbyInfo.ClientInSlot(sl.Key) == null) ||
-					(!orderManager.LobbyInfo.GlobalSettings.EnableSingleplayer && orderManager.LobbyInfo.NonBotPlayers.Count() < 2);
+					(!orderManager.LobbyInfo.GlobalSettings.EnableSingleplayer && orderManager.LobbyInfo.NonBotPlayers.Count() < 2) ||
+					map.SpawnPoints.Length - orderManager.LobbyInfo.Clients.Where(c => !c.IsObserver).Count() - orderManager.LobbyInfo.DisabledSpawns.Count < 0;
```
```suggestion
					map.SpawnPoints.Length - orderManager.LobbyInfo.Clients.Count(c => !c.IsObserver) - orderManager.LobbyInfo.DisabledSpawns.Count < 0;
```

teinarss commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -116,6 +117,10 @@ static void CheckAutoStart(S server)
 				return;
 
+			// Can't have insufficient spawns
+			if (server.Map.SpawnPoints.Length - server.LobbyInfo.Clients.Where(c => !c.IsObserver).Count() - server.LobbyInfo.DisabledSpawns.Count < 0)
```
```suggestion
			if (server.Map.SpawnPoints.Length - server.LobbyInfo.Clients.Count(c => !c.IsObserver) - server.LobbyInfo.DisabledSpawns.Count < 0)
```

teinarss commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -148,6 +153,12 @@ static bool StartGame(S server, Connection conn, Session.Client client, string s
 				return true;
 			}
 
+			if (server.Map.SpawnPoints.Length - server.LobbyInfo.Clients.Where(c => !c.IsObserver).Count() - server.LobbyInfo.DisabledSpawns.Count < 0)
```
```suggestion
			if (server.Map.SpawnPoints.Length - server.LobbyInfo.Clients.Count(c => !c.IsObserver) - server.LobbyInfo.DisabledSpawns.Coun
```

ocdi commented:
I rebased on bleed

teinarss commented:
Just one thought, maybe reset the disabled spawn points on a map change?

dragunoff commented:
I think a properly sized multiplication sign ["×"](https://unicode-table.com/en/00D7/) looks better than an "X". And overall it may be more appropriate to do this with a sprite (but this shouldn't be a blocker for this PR).

![image](https://user-images.githubusercontent.com/1355810/91053103-062fce80-e62b-11ea-932e-20c58aedbf4d.png)

pchote commented:
We're (hopefully) close to branching a new playtest series, so we should either make sure that this is polished enough to ship here, or hold off merging it until after the branch.

I agree that this should be done with a sprite version of the spawn point artwork.

ocdi commented:
100% agree that a sprite should be used, I'm not familiar enough with that side to implement that. I will give the Multiplication sign a go. :-)

ocdi commented:
I had a look into using the multiplication x and it was actually more work than starting to modify the sprite to introduce the x

This seems to be the chrome-2x.png file and I drew the lines with 4px thickness. Do you think this has enough contrast here or should it be a brighter red? Wanted opinions before I spent the time going through all the base mods to add this sprite.

![image](https://user-images.githubusercontent.com/2775804/91228735-1a2f1980-e76c-11ea-9313-38e82df1cff8.png)

I also did @teinarss 's suggestion of resetting disabled spawns upon map change.

pchote commented:
What if we hid the spawn letter and used a white X through the background?

Also, please note that the TD chrome*.png and other mod glyph*.png are generated from the svg files in https://github.com/OpenRA/ArtSrc/, and can't be edited manually!

ocdi commented:
I had a quick go at editing the svg using inkscape, I did your suggesion of doing the white x and removing the x. The chrome-3x.png appears to be 288dpi export but expanded to 4096x2048 - is that correct or is the 3x image not actually used? This is the image comparison in my git client

![image](https://user-images.githubusercontent.com/2775804/91234525-c675fd80-e776-11ea-8d45-122217434b20.png)

Also I used Inkscape 1.0 to edit the svg and the diff is extremely large. The export appear to be the same except for my added sprite.

![image](https://user-images.githubusercontent.com/2775804/91234818-75b2d480-e777-11ea-8d29-3f6a1e1ee99d.png)

ocdi commented:
d2k ![image](https://user-images.githubusercontent.com/2775804/91288221-1636e200-e7d4-11ea-9927-2508dbd86a47.png)

ra 
![image](https://user-images.githubusercontent.com/2775804/91289270-7c703480-e7d5-11ea-8bc5-f35b62a25954.png)

When testing RA I accidentally forgot to update the mod manifest, which got me thinking, what about other mods who might not define a sprite. In this case I am falling back to unclaimed if there is no closed spawn sprite, with the letter gone it looks acceptible, but open to other suggestions here.
![image](https://user-images.githubusercontent.com/2775804/91289553-d96bea80-e7d5-11ea-9752-a9121c6f83bf.png)

ts 
![image](https://user-images.githubusercontent.com/2775804/91290060-80508680-e7d6-11ea-878d-ab3870c65d44.png)

I also exported 288dpi for the 3x images, but manually expanded the canvas to match the original images.

dragunoff commented:
> 
> I also exported 288dpi for the 3x images, but manually expanded the canvas to match the original images.

@ocdi This is correct. The sprite needs to be manually expanded because sprite dimensions have to be a power of two. The 3x sprite is used on high dpi displays. 

The graphic looks good however it's a little larger than the spawn point graphic. I guess this was a deliberate decision so that it covers the graphic below but perhaps it looks a bit odd:
![image](https://user-images.githubusercontent.com/1355810/91295101-37330f00-e7a3-11ea-910d-b2c58f6297e5.png)

Could we pull it off by replacing the spawn point graphic and removing the letter?

ocdi commented:
Actually the X is part of the sprite, I removed the text output completely. I think my "mistake" was picking the selected spawn position spirite as the base that I copied which is the slightly larger compared to the unselected spawn one.

I can do it again with the smaller base if you like. Out of interest, for your example above if you choose spawn points for players I imagine it should look larger.

dragunoff commented:
> Actually the X is part of the sprite, I removed the text output completely. I think my "mistake" was picking the selected spawn position spirite as the base that I copied which is the slightly larger compared to the unselected spawn one.
> 
> I can do it again with the smaller base if you like. Out of interest, for your example above if you choose spawn points for players I imagine it should look larger.

Ah, I see now. Then I think it's correct that it's based on the occupied spawn point graphic. Because a disabled point is no longer available (occupied in a sense).

![image](https://user-images.githubusercontent.com/1355810/91306487-3dca8200-e7b5-11ea-8359-1d5b6eae0a96.png)

However I noticed that in RA there seems to be an overlap (visible on your screenshot as well).

![image](https://user-images.githubusercontent.com/1355810/91306990-0ad4be00-e7b6-11ea-963b-f64f8269ec13.png)

pchote commented:
I'd like to see a version (a mockup is fine) using the unoccupied sprite. IMO it would probably look better.

The reason why the occupied marker is larger is that, historically, the engine was not able to draw coloured circles, only squares. The larger ring was needed to cover the corners of the square.

ocdi commented:
Just heading to bed, will make new version in the morning

ocdi commented:
![image](https://user-images.githubusercontent.com/2775804/91362862-b0cc0b00-e83e-11ea-8bcc-be2a0378bf71.png)

This is the new version. I discovered there is a circle being drawn before the sprite, which is what that weird black was. I am suppressing that for closed slots now. I think the smaller images looks better so will do the same for the other mods.

ocdi commented:
OK Did the new sprites for the mods.

ocdi commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -766,16 +766,25 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
 		static bool ToggleSpawn(S server, Connection conn, Session.Client client, string s)
 		{
 			if (!client.IsAdmin)
-				return false;
```
The behaviour I am seeing is false means the command was not found, true means handled even though it might not do anything. I figured returning a message here is better than "command not found".

ocdi commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -766,16 +766,25 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
 
-			if (server.LobbyInfo.DisabledSpawns.Contains(spawn))
-				server.LobbyInfo.DisabledSpawns.Remove(spawn);
+			if (server.LobbyInfo.Clients.Any(cc => cc.SpawnPoint == spawnPoint))
```
I copied the relevant bit of this check from the `Spawn` method; the variable used there was `spawnPoint` so I adjusted to the same here.

ocdi commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -766,16 +766,25 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
-				return false;
+			{
+				server.SendOrderTo(conn, "Message", "Only admins can disable spawn points.");
+				return true;
```
The behaviour I am seeing is false means the command was not found, true means handled even though it might not do anything. I figured returning a message here is better than "command not found".

ocdi commented:
Also did some more testing, the last couple were bug fixes introduced related to some feedback and there was a race condition where you rapidly left then right click on a spawn and before the lobby syncs it would issue the toggle spawn command allowing you to disable a chosen spawn. A very edge case but is now handled better. I put some comments on those commits explaining the changes.

I will squash everything once you guys are happy with these most recent changes.

pchote commented:
A few thoughts from a quick test ingame:
* Disabled spawns should show a "Disabled spawn" tooltip.
* IMO the grey background on the disabled TD spawn points looks odd. Why not the original black?
* Contrast doesn't feel great between the white X and grey background on RA etc - what if we had a black stroke that matches the spawn letters?
* Its a bit awkward to have right click disable an empty spawn, but do nothing on an occupied spawn. It would be much better if right click on an occupied spawn could remove the player.
* It feels a bit odd to have no feedback when you are trying to remove more spawns than there are currently players. I wonder whether we should add the standard click sound when selecting a spawn on the minimap, and then use the click-disabled sound on failure?
* Serious bonus points if you can also identify and fix whatever is causing the spawn letters to be drawn off-center in the markers. This has mildly bugged me for years (but never quite at the level to look into it myself).

I haven't had time to look at the code yet.

ocdi commented:
1. Disabled spawn tooltip was broken by me skipping rendering the text - I have a commit that fixes that
2. Copy/paste of the sprite from the other versions. :-)
3. Agree - I took your white suggestion from cnc and applied universally but that didn't work with the grey. I will restore the original darker background and use the white for TD and make it black with the grey for others.
4. I have a commit that will forcibly close spawns
5. I haven't discovered the code to play audio yet, but I changed it to show a message like for other error states, which had the bonus of playing an audio tone
6. I had a look but it is a weird one, there is an offset being used but adjusting fixes some but not others. I suspect there are some characters that are an odd number of pixels which given the font sizes would lead to the off-center issue.

cnc - *much* better
![image](https://user-images.githubusercontent.com/2775804/91625121-1f000180-e9e8-11ea-981f-1f8a8b205fc7.png)

ra - I matched the border colour
![image](https://user-images.githubusercontent.com/2775804/91625170-828a2f00-e9e8-11ea-90a5-cda27154c9b7.png)

d2k
![image](https://user-images.githubusercontent.com/2775804/91625257-2a9ff800-e9e9-11ea-8667-688587d7cd7f.png)

ts
![image](https://user-images.githubusercontent.com/2775804/91625295-82d6fa00-e9e9-11ea-98d4-dcc5957c3d79.png)

Will squash and rebase once you are happy with the final code.

ocdi commented:
I did some more looking into your bonus question and stepped into the SpriteFont code for the TinyBold to inspect the measurements. It appears the A character ends up as 11x11 pixels, which is clearly the cause of the odd off-center behaviour. I'm not sure it is possible to ever be 100% correct with that. Instead I figured I would try making the font larger, bumping to 12 looks like this:

![image](https://user-images.githubusercontent.com/2775804/91625717-0ba36500-e9ed-11ea-9491-169a3f9cbe1f.png)

The width becomes 13x14, which doesn't solve the offset, I think the visual appearance of the offset is less.
The 'B' here is 11x14, which does look slightly to the left, but that could be visually biased slightly due to the straight edge and the rounder bumps to the right.

'C' ends up as 13x15, which again is still an odd number of pixels.

But I think this looks much more centered overall:
![image](https://user-images.githubusercontent.com/2775804/91625848-e82cea00-e9ed-11ea-9256-e3b5dbb8e03d.png)

ocdi commented:
I tried applying the font change to the other mods and that made it worse over there, then I realised that cnc was the only one that appears to have issues. I adjusted the artwork to 18px for the spawn circles rather than the 19px as is current in the artwork. The PR I just created adjusts the offset for the current sprites, but I can commit the artwork changes if you think that looks better:

![image](https://user-images.githubusercontent.com/2775804/91626609-f2ea7d80-e9f3-11ea-9040-e6d54ef4dabc.png)

The PR has an original comparison shot vs the one above with smaller circles.

Open to suggestions as to the best change here.

pchote commented:
Since we're changing the artwork anyway, it would make sense to fix that here too. I suspect that some of the problems may be due to point-vs-pixel alignment with HiDPI rendering, which is a bit complicated and unfair to scope creep onto a newcomer!

So I think it would make sense to change the size so we can safely assume that all mods use even-sized spawn images, and then leave further fixes to a followup PR.

pchote commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -69,6 +71,9 @@ public static Session Deserialize(string data)
 							session.Slots.Add(s.PlayerReference, s);
 							break;
+						case "DisabledSpawns":
+							session.DisabledSpawns = Exts.ParseCommaIntegerArrayInvariant(node.Value.Value).ToList();
```
We already have code for this!

Use `FieldLoader.GetValue<int[]>("DisabledSpawns", node.Value.Value).ToList()` then you can remove `ParseCommaIntegerArrayInvariant`.

pchote replied:
Note that `FieldLoader` doesn't support lists - if you feel like making some deeper changes to make this a bit cleaner you could either redefine `DisabledSpawns` to an array, or look into adding List support to FieldLoader (warning: this probably won't be easy!)

ocdi replied:
Actually turns out `FieldLoader` and `FieldSaver` was fairly easy to update to support `List<T>`. I started writing the code next to the `HashSet<T>` and then realised my new code was identical, so just altered the type check to be either `HashSet<T> || List<T>`. It works.

pchote commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -267,7 +272,10 @@ public string OptionOrDefault(string id, string def)
-			var sessionData = new List<MiniYamlNode>();
+			var sessionData = new List<MiniYamlNode>()
+			{
+				new MiniYamlNode("DisabledSpawns", string.Join(",", DisabledSpawns))
```
Then this can change to `FieldSaver.FormatValue(DisabledSpawns.ToArray())`.

pchote commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -115,6 +116,10 @@ static void CheckAutoStart(S server)
 				return;
 
+			// Can't have insufficient spawns
+			if (server.Map.SpawnPoints.Length - server.LobbyInfo.Clients.Count(c => !c.IsObserver) - server.LobbyInfo.DisabledSpawns.Count < 0)
```
This line is a bit long/unclear. Can we change this to something like
```csharp
var availableSpawnCount = server.Map.SpawnPoints.Length - server.LobbyInfo.DisabledSpawns.Count;
if (availableSpawnCount < server.LobbyInfo.Clients.Count(c => !c.IsObserver))
	return;
```
?

pchote commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -147,6 +152,12 @@ static bool StartGame(S server, Connection conn, Session.Client client, string s
 				return true;
 			}
 
+			if (server.Map.SpawnPoints.Length - server.LobbyInfo.Clients.Count(c => !c.IsObserver) - server.LobbyInfo.DisabledSpawns.Count < 0)
```
And then likewise here.

pchote commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -147,6 +152,12 @@ static bool StartGame(S server, Connection conn, Session.Client client, string s
 
+			if (server.Map.SpawnPoints.Length - server.LobbyInfo.Clients.Count(c => !c.IsObserver) - server.LobbyInfo.DisabledSpawns.Count < 0)
+			{
+				server.SendOrderTo(conn, "Message", "Unable to start the game, too many spawn points are disabled for the number of players.");
```
Maybe "Unable to start the game until more spawn points are enabled"?

pchote commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -750,6 +763,45 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
+				// check we have enough to toggle
+				if (server.Map.SpawnPoints.Length - server.LobbyInfo.Clients.Count(c => !c.IsObserver) - server.LobbyInfo.DisabledSpawns.Count < (existingClient != null ? 1 : 0))
+				{
+					server.SendOrderTo(conn, "Message", "Unable to disable as there would be insufficient spawn locations for all players.");
```
IMO the wording here is a bit awkward, but I don't have any immediate suggestions that would be better. Maybe someone else has ideas?

ocdi replied:
Agree, perhaps:

* Unable to disable spawn location, players need a start location.
* All players will require a start location so spawn is still open.

pchote replied:
There is another edge case to worry about here: the admin can disable the maximum number of allowed spawns, but there is nothing stopping a spectator from switching into an empty player slot, now causing there to be less free spawns than players. That player can ready up, but nothing will happen (not even a warning to the chat about insufficient free slots).

IMO trying to deal with that edge case and coming up with sensible error messages for the different cases is more complexity than it is worth. It would be better if we could instead allow the admin to disable as many spawns as they want, and have a single clear error message that comes up when all the players are ready explaining that the game can't start until the situation is resolved.

ocdi replied:
This is mostly the behaviour now - you can disable as many as you like, the only problem piece missing is the lack of notification why the game can't start currently.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -368,7 +368,8 @@ void ConnectionStateChanged(OrderManager om)
 					orderManager.LobbyInfo.Slots.Any(sl => sl.Value.Required && orderManager.LobbyInfo.ClientInSlot(sl.Key) == null) ||
-					(!orderManager.LobbyInfo.GlobalSettings.EnableSingleplayer && orderManager.LobbyInfo.NonBotPlayers.Count() < 2);
+					(!orderManager.LobbyInfo.GlobalSettings.EnableSingleplayer && orderManager.LobbyInfo.NonBotPlayers.Count() < 2) ||
+					map.SpawnPoints.Length - orderManager.LobbyInfo.Clients.Count(c => !c.IsObserver) - orderManager.LobbyInfo.DisabledSpawns.Count < 0;
```
This condition will rarely change, so it feels a bit awkward to be evaluating this at the render rate.

It would be better if we could treat this like `disableTeamChat`: define an `insufficientPlayerSpawns` variable that is updated in `UpdatePlayerList` (which is automatically called whenever the lobbyinfo updates)

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -230,9 +230,15 @@ public SlotDropDownOption(string title, string order, Func<bool> selected)
 		{
 			var spawns = preview.SpawnPoints;
-			return lobbyInfo.Clients
+			var occupants = lobbyInfo.Clients
```
Mixing LINQ and manual enumeration isn't great.

It would be better if we either stick completely with LINQ, something like:
```csharp
return lobbyInfo.Clients
	.Where(c => (c.SpawnPoint - 1 >= 0) && (c.SpawnPoint - 1 < spawns.Length))
	.Select(c => new SpawnOccupant(c))
	.Concat(lobbyInfo.DisabledSpawns
		.Where(d => d - 1 < spawns.Length)
		.Select(d => new SpawnOccupant(d)))
	.ToDictionary(c => spawns[c.SpawnPoint - 1], c => c);
```

or define the empty dictionary and then populate it using two `foreach` loops.

pchote reviewed (commented):
I ran out of time to finish reviewing the UI logic, but here's where I got to with the rest of the code:

ocdi commented:
@pchote thanks for assisting cleaning up the artwork changes. I have pulled your changes into my branch and rebased, squashed. I left your sprite change commit separate but can squash that also.

pchote commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -750,6 +765,46 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
+
+			if (existingClient != null)
+			{
+				existingClient.SpawnPoint = 0;
```
We should also set
```csharp
if (existingClient.State == Session.ClientState.Ready)
    existingClient.State = Session.ClientState.NotReady;
```
IMO.

pchote commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -750,6 +765,46 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
 			return true;
 		}
 
+		static bool ToggleSpawn(S server, Connection conn, Session.Client client, string s)
```
IMO it is still annoyingly awkward to go straight from an active spawn containing a player to disabled. It would be much more usable if we could refocus this as a "clear spawn" feature, where clearing an active spawn removes the player, and clearing an empty spawn disables it.

Something like this:
```csharp
static bool ClearPlayerSpawn(S server, Connection conn, Session.Client client, string s)
{
	var spawnPoint = Exts.ParseIntegerInvariant(s);
	if (spawnPoint == 0)
		return true;

	var existingClient = server.LobbyInfo.Clients.FirstOrDefault(cc => cc.SpawnPoint == spawnPoint);
	if (client != existingClient && !client.IsAdmin)
	{
		server.SendOrderTo(conn, "Message", "Only admins can clear spawn points.");
		return true;
	}

	// Clearing a selected spawn point removes the player
	if (existingClient != null)
	{
		existingClient.SpawnPoint = 0;
		if (existingClient.State == Session.ClientState.Ready)
    			existingClient.State = Session.ClientState.NotReady;

		server.SyncLobbyClients();
		return true;
	}

	// Clearing an empty spawn point prevents it from being selected
	// Clearing a disabled spawn restores it for use
	if (!server.LobbyInfo.DisabledSpawns.Contains(spawnPoint))
		server.LobbyInfo.DisabledSpawns.Add(spawnPoint);
	else
		server.LobbyInfo.DisabledSpawns.Remove(spawnPoint);

	server.SyncLobbyInfo();
	return true;
}
```

and then in LobbyUtils.cs:
```csharp
static void ClearPlayerSpawn(OrderManager orderManager, MapPreviewWidget mapPreview, MapPreview preview, MouseInput mi)
{
	var selectedSpawn = DetermineSelectedSpawn(mapPreview, preview, mi);
	if (Game.IsHost || orderManager.LobbyInfo.Clients.FirstOrDefault(cc => cc.SpawnPoint == selectedSpawn) == orderManager.LocalClient)
		orderManager.IssueOrder(Order.Command("clear_spawn {0}".F(selectedSpawn)));
}
```

ocdi replied:
I removed the `Game.IsHost` check in the `SelectSpawnPoint` method, following from your logic in the `ClearPlayerSpawn` method, this allows players to vacate a spawn using RMB themselves. The alert of insufficient spawns and preventing disabling too many is removed, but that solves the awkward wording issue, which is handled by the disabled Start Game button.

pchote commented on OpenRA.Game/FieldLoader.cs:
```diff
@@ -435,7 +435,7 @@ public static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, M
 				return ret;
 			}
-			else if (fieldType.IsGenericType && fieldType.GetGenericTypeDefinition() == typeof(HashSet<>))
+			else if (fieldType.IsGenericType && (fieldType.GetGenericTypeDefinition() == typeof(HashSet<>) || fieldType.GetGenericTypeDefinition() == typeof(List<>)))
```
Can we please pull these changes out into their own commit "Add FieldLoader/Saver support for List<T>"?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -245,27 +249,46 @@ public SlotDropDownOption(string title, string order, Func<bool> selected)
+				ToggleSpawn(orderManager, mapPreview, preview, mi);
+		}
 
+		private static void SelectPlayerSpawn(OrderManager orderManager, MapPreviewWidget mapPreview, MapPreview preview, MouseInput mi)
```
OpenRA code style is to omit explicit `private`s, since this is already the default.

pchote commented:
I have some more review comments to come, but some of them are tied in with existing bad code that interferes with doing things better here, so I will probably split out a separate PR to fix those first.

pchote commented:
Looking at the UI side of this in detail has revealed a can of worms:
 * The lobby UI (predating this PR) is doing some weird inefficient things
    &rarr; I filed #18667 to try and improve this (see [1](https://github.com/pchote/OpenRA/commit/13011ad6aab46208a2f144832576f41f1d7b1fae#diff-455cbcee4229a4abaf8491a63e23cdc4R801-R804) and [2](https://github.com/pchote/OpenRA/commit/13011ad6aab46208a2f144832576f41f1d7b1fae#diff-33b127a7c4c3a68db1faebc6e256b27aR203-R219) for my suggestions on how to fix the rebase conflicts here)
 * I don't think it is a good idea to implement "spawn disabled" as special case of "spawn occupied". IMO this isn't expected, and may lead to maintenance issues in the future.
    &rarr; I prototyped https://github.com/pchote/OpenRA/commit/9ed09a406a545d48d445c72446f33fd759889bb8  as a suggested alternative.
 * As commented in the commit in my last point, we will need to update the master server protocol and replay format to display disabled spawns in the server and replay browsers.
 * The map format allows specific spawn points to be reserved for specific player slots, but the logic behind this is a real mess. We will need to make sure that the server prevents admins from disabling reserved spawns.
   &rarr; I filed #18672 to try and improve this. This PR will need to be rebased and updated to check for reserved spawns once that has been merged.

pchote commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -26,6 +26,8 @@ public class Session
 		// Keyed by the PlayerReference id that the slot corresponds to
 		public Dictionary<string, Slot> Slots = new Dictionary<string, Slot>();
 
+		public List<int> DisabledSpawns = new List<int>();
```
I've been flip-flopping around on names, but with #18672 settled on using "SpawnPoint" for these integer labels. Can we please rename this and the methods that touch it to use "SpawnPoint" too?

ocdi replied:
I have pushed a commit that does this - I hope I have interpreted your wishes correctly :-)

pchote reviewed (commented):
A couple of smaller style comments to wrap up this review:

ocdi commented:
I took your changes @pchote and rebased on the separate PR I did and brought it all together. Still seems to be working as originally intended but with your reworked occupancy changes. I also prepared a separate branch that was rebased on your work with this being the key bit to restore the disabled spawns after those changes

https://github.com/ocdi/OpenRA/commit/ffb91c66a1d2ab6bca8578c95569457e982c470e

Hopefully that looks good to you.

ocdi commented:
Rebased

pchote commented:
I'll give this another review pass once my two spawn point PRs have been merged. I'm reasonably confident that the plumbing should be sound once they are merged and this rebased, so we can hopefully get this merged soon after.

In the meantime it would be good if you could work towards the missing replay and server list display.

Replays should be simple, we just need to add the disabled spawn list to the `GameInformation` class, and the underlying replay logic will automatically save and load it thanks to your FieldLoader fix.

Likewise with `GameServer` for the server list. You should be able to test this using two clients and the LAN game broadcasts, and internet games will start working once we update the master server backend to match.

ocdi commented:
@pchote I rebased this again and implemented the replay and server browser TODOs you had left behind.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/MapPreviewLogic.cs:
```diff
@@ -172,12 +172,13 @@ public override void Tick()
 
 		void SetupWidgets(Widget parent, Func<MapPreview> getMap,
-			Action<MapPreviewWidget, MapPreview, MouseInput> onMouseDown, Func<Dictionary<int, SpawnOccupant>> getSpawnOccupants, bool showUnoccupiedSpawnpoints)
+				Action<MapPreviewWidget, MapPreview, MouseInput> onMouseDown, Func<Dictionary<int, SpawnOccupant>> getSpawnOccupants, Func<List<int>> getDisabledSpawnPoints, bool showUnoccupiedSpawnpoints)
```
Style nit: This change doesn't look correct.

abcdefg30 commented on OpenRA.Game/GameInformation.cs:
```diff
@@ -35,6 +35,9 @@ public class GameInformation
 		public IList<Player> Players { get; private set; }
+
+		public List<int> DisabledSpawnPoints;
+
```
Style nit: Can we remove the empty lines above and below the new list again to be consistent with the other properties?

abcdefg30 reviewed (changes requested):
I'm no longer getting tooltips for the map preview.

ocdi commented:
@abcdefg30 I restored the tooltips and fixed your style nits.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ServerListLogic.cs:
```diff
@@ -427,11 +427,7 @@ void SelectServer(GameServer server)
-				// This requires further changes to MasterServerPinger and the master server backend
-				var noSpawns = new List<int>();
-				mapPreview.DisabledSpawnPoints = () => noSpawns;
+				mapPreview.DisabledSpawnPoints = () => server.DisabledSpawnPoints ?? new List<int>();
```
This will create a new list every time the UI is rendered if `server.DisabledSpawnPoints` is null!

We could either make sure that `server.DisabledSpawnPoints` like I suggested ~above~ below for replays, otherwise pull out a `var disabledSpawns = server.DisabledSpawnPoints ?? new List<int>();` and return that like we do for `occupants`.

pchote commented on OpenRA.Game/GameInformation.cs:
```diff
@@ -35,6 +35,7 @@ public class GameInformation
 		/// <summary>Gets the game's duration, from the time the game started until the replay recording stopped.</summary>
 		public TimeSpan Duration { get { return EndTimeUtc > StartTimeUtc ? EndTimeUtc - StartTimeUtc : TimeSpan.Zero; } }
 		public IList<Player> Players { get; private set; }
+		public List<int> DisabledSpawnPoints;
```
It would probably be cleaner to default this to an empty list rather than null, then we can avoid having to manually fall back to an empty list in `ReplayBrowserLogic`. What do you think?

pchote commented on OpenRA.Game/Network/GameServer.cs:
```diff
@@ -226,6 +229,7 @@ public GameServer(Server.Server server)
 			Protected = !string.IsNullOrEmpty(server.Settings.Password);
 			Authentication = server.Settings.RequireAuthentication || server.Settings.ProfileIDWhitelist.Any();
 			Clients = server.LobbyInfo.Clients.Select(c => new GameClient(c)).ToArray();
+			DisabledSpawnPoints = server.LobbyInfo.DisabledSpawnPoints;
```
We will need to clone the list to prevent server changes from mutating this record of the state. Use `.ToList()` or, better (because it is more immutable), define `DisabledSpawnPoints` as an array and use `.ToArray()`

pchote commented on mods/cnc/chrome.yaml:
```diff
@@ -306,8 +306,9 @@ music:
-		spawn-claimed: 777, 443, 19, 19
+		spawn-claimed: 744, 215, 18, 18
+		spawn-unclaimed: 744, 234, 18, 18
+		spawn-closed: 744, 253, 18, 18
```
Can we please change this to `spawn-disabled` to complete the (re)naming?

pchote commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -811,6 +826,41 @@ static bool Team(S server, Connection conn, Session.Client client, string s)
+				server.SyncLobbyClients();
+				return true;
+			}
+
```
I was expecting to see a check here to make sure that the spawn wasn't locked to a player slot.

Test this by adding
```
LockSpawn: True
Spawn: 1
```
under `PlayerReference@Multi0:` in map.yaml for Fort Lonestar (our favorite map for exposing bugs in lobby/server PRs) or another map of your choice.

pchote reviewed (commented):
A few small comments from reading the code. I bet you weren't expecting this PR to turn into such an ordeal when you first started it 🙂 

I want to hold off testing this until #18696 has been merged, but this should hopefully be robust enough now that I'm not expecting to find any new major problems.

ocdi commented:
> I bet you weren't expecting this PR to turn into such an ordeal when you first started it 🙂

lol it has certainly been more involved than I expected; certainly learned a lot. :+1: I had picked what I thought was a simple/quick issue that was good to learn how the game works. It has been fun and from where I started on this it has improved significantly.

pchote commented:
#18696 has finally been merged. If you're able to rebase this this weekend i'll do my best to review it ASAP. Otherwise, i'm going to be away until at least the 12th and will take a look once I get back.

ocdi commented:
I rebased on bleed and surprisingly there were no conflicts, which was a first. 👍

pchote commented on OpenRA.Mods.Common/Traits/World/MPStartLocations.cs:
```diff
@@ -130,7 +130,7 @@ void INotifyCreated.Created(Actor self)
 
 			// Initialize the list of unoccupied spawn points for AssignSpawnLocations to pick from
-			availableSpawnPoints = Enumerable.Range(1, spawnLocations.Length).ToList();
+			availableSpawnPoints = Enumerable.Range(1, spawnLocations.Length).Except(self.World.LobbyInfo.DisabledSpawnPoints).ToList();
```
> I rebased on bleed and surprisingly there were no conflicts, which was a first. 👍

Not quite!

It wasn't an explicit conflict, but we will now also need to make a matching change in `object IAssignSpawnPointsInfo.InitializeState(MapPreview map, Session lobbyInfo)` in `MPStartLocationsInfo` above.

FWIW this expected conflict was why I was waiting for #18696 to be merged.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -229,29 +229,48 @@ public SlotDropDownOption(string title, string order, Func<bool> selected)
 
 		public static void SelectSpawnPoint(OrderManager orderManager, MapPreviewWidget mapPreview, MapPreview preview, MouseInput mi)
 		{
```
We will need a ready check here:
```csharp
if (orderManager.LocalClient.State == Session.ClientState.Ready)
	return;
```

to avoid a bug:
* Create a lobby with more spawns disabled than there are players (so the game doesn't autostart at the next step)
* Check the ready checkbox next to your name in the player list
* Right click on an enabled spawn

Expected behaviour: Nothing happens
Actual behaviour: A pair of server error messages related to setting state while ready

The check in `SelectPlayerSpawnPoint` can then be removed, and the observer part of that was wrong!

pchote commented on OpenRA.Game/Network/GameServer.cs:
```diff
@@ -132,6 +132,9 @@ public class GameServer
 		public readonly GameClient[] Clients;
 
+		/// <summary>The list of spawnpoints that are disabled for this game</summary>
+		public readonly int[] DisabledSpawnPoints;
```
Please initialize this as `public readonly int[] DisabledSpawnPoints = { };`. Otherwise, the game crashes when selecting a server in the multiplayer browser.

pchote reviewed (commented):
A couple new issues found while testing.

Can you please also squash the fixup commits again ready to merge? It would be easiest and IMO best to have just one commit with a message like "Right click lobby spawns to disable or remove players."

ocdi commented:
OK I have addressed those final 3 issues and squashed to the single commit as requested.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -229,29 +229,48 @@ public SlotDropDownOption(string title, string order, Func<bool> selected)
+				orderManager.IssueOrder(Order.Command("clear_spawn {0}".F(selectedSpawn)));
+		}
+
+		private static int DetermineSelectedSpawnPoint(MapPreviewWidget mapPreview, MapPreview preview, MouseInput mi)
```
Style nit: Remove `private` here.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -229,29 +229,48 @@ public SlotDropDownOption(string title, string order, Func<bool> selected)
 		}
 
-		private static void SetSpawnPoint(OrderManager orderManager, Session.Client playerToMove, int selectedSpawn)
+		private static void SetSpawnPoint(OrderManager orderManager, Session.Client playerToMove, int selectedSpawnPoint)
```
Same here.

abcdefg30 reviewed (commented):
Lgtm on a cursory look otherwise.

ocdi commented:
@abcdefg30 fixed those nit's, to be fair `SetSpawnPoint` had the private modifier previously, so at least while touching it I am cleaning it up. :-)

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/adfd9c8b796be8d4480efa1d6a1f5a868ab3615b...43ad1c292ed904a828687fa895624c95510a5ce5)
